File: hw6/Iterator.cpp
# include "Iterator.h"
# include <iterator>
namespace jwan_{

	template<class E, template <class... > class Container>
	Iterator<E,Container>::Iterator(){

	}

	template<class E, template <class... > class Container>
	Iterator<E,Container>::Iterator(Container< E>& o){
		int i =0;
		it = new E[o.size()] ;
		auto temp = o.begin();
		for( i = 0 ; i <o.size() ; i++ ){
			it[i] = *temp++ ; 
		}
		end = new E ;
		end = &it[i] ;
	}

	template<class E, template <class... > class Container>
	Iterator<E,Container>::Iterator(const Iterator& o){
		it = new E ;
		it = o.it ;
		end = new E ;
		end = o.end ;
	}	
	
	template<class E, template <class... > class Container>
	void Iterator<E,Container>::print(){
		std::cout << *it ;
	}

	template<class E, template <class... > class Container>
	bool Iterator<E,Container>::hasnext(){
		if ( it < end  ){
			return true ;
		}
		return false ;
	}

	template<class E, template <class... > class Container>
	void Iterator<E,Container>::next(){
		if (hasnext()){
			it = it+1 ;
		}  
	}

	template<class E, template <class... > class Container>
	void Iterator<E,Container>::remove(){
		if ( it < end ){
			end = end - 2 ;
		}
	}

	template<class E, template <class... > class Container>
	Iterator<E,Container>& Iterator<E,Container>::operator = (const Iterator& o){
		if (o.it != it  ){
			it = new E ;
			it = o.it ;
			end = new E ;
			end = o.end ;
		}
	}

	template<class E, template <class... > class Container>
	E& Iterator<E,Container>::operator *(){
		return *it ;
	}
}

File: hw8/iterator.java

import java.lang.IllegalArgumentException;
@SuppressWarnings("unchecked")
public class iterator<E> {

	public iterator(){
		
	}
	public iterator(E[] other){
		arr = other ;

		for(int i =0 ; i < other.length ; i++){
			arr[i] = other[i] ;
		}
		index = 0 ;
	}

	public E next(){
		try{
			if(index < arr.length){
				index++;
				return arr[index-1];
			}
			else {
				throw new IllegalAccessException("no next element in itration\n");
			}
			
		}catch(IllegalAccessException e){
			System.out.printf(e.getMessage());
		}
		return null;

	}
	public boolean hasNext(){
		if(index < arr.length){
			return true ;
		}
		return false ;
	}
	public void remove(){
		E[] temp = arr ;
		int new_size =  arr.length-1 ;
		arr = (E[])new Object[new_size];
		for(int i = 0 ; i < arr.length ; i++){
			arr[i] = temp[i];
		}
	}


	private E[] arr ;
	private int index;

}